### PR TITLE
add package tbb as well as some fixes to previous packages

### PR DESCRIPTION
**Pull request by xq114**
explanation for openexr: package openexr is composed of two parts and required by some other packages on both static and shared library.

### CODE REVIEW

**Comment by waruqi** [comment]
Please merge into a single commit to reduce redundant commit history.

```console
git reset --soft c7739b3
git commit -a -m "add tbb and fix some packages"
git push origin master --force
```

**Comment by waruqi** [comment]
Why are there repeated redundant commits? Some of these commits have been merged before. Please keep a single commit, so we can pulling repo as quickly as possible when install packages.

**Comment by xq114** [comment]
A regression error occurs. I have to wait until it is resolved, see [https://github.com/AcademySoftwareFoundation/openexr/issues/856](https://github.com/AcademySoftwareFoundation/openexr/issues/856). Should I close this pr or keep it open?

**Comment by waruqi** [comment]
> A regression error occurs. I have to wait until it is resolved, see [AcademySoftwareFoundation/openexr#856](https://github.com/AcademySoftwareFoundation/openexr/issues/856). Should I close this pr or keep it open?

Why build static/shared libraries by default? add_requires only needs to install static libraries by default, unless we pass the shared = true parameter to add_requires to tell xmake to install dynamic libraries. Usually there is no need to have static/shared libraries at the same time.

You can remove build_both config or set it false by default

File: packages/o/openexr/xmake.lua [hunk 1]
```diff
@@ -11,19 +11,26 @@ package("openexr")
...
+            table.insert(configs, "-DOPENEXR_BUILD_BOTH_STATIC_SHARED=OFF")
+            table.insert(configs, "-DILMBASE_BUILD_BOTH_STATIC_SHARED=OFF")
         end
         table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
         table.insert(configs, "-DPYILMBASE_ENABLE=" .. "OFF")
         import("package.tools.cmake").install(package, configs)
+        for _, name in ipairs(os.dirs(os.curdir() .. "/build_*")) do
+            os.cp(name .. "/install/bin", package:installdir())
+            break
+        end
```
Comment (waruqi) [inline packages/o/openexr/xmake.lua:33]:
```console
import("package.tools.cmake").install(package, configs, {buildir = "build"})
os.cp("build/install/bin", package:installdir())
```

File: packages/o/openexr/xmake.lua [hunk 2]
```diff
@@ -11,12 +11,23 @@ package("openexr")
     add_deps("cmake")
     add_deps("zlib")
 
-    on_install("macosx", "linux", "windows", "mingw", function (package)
-        local configs = {"-DBUILD_TESTING=OFF", "-DBUILD_OPENEXR_EXAMPLES=OFF"}
-        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+    add_configs("build_both", {description = "Build both static library and shared library.", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/o/openexr/xmake.lua:14]: 这里为什么需要同时安装动态库和静态库的选项配置呢？

如果除非必要，请去掉 build_both 选项，我们应该尽可能退化成 仅仅只安装一个版本，要么动态库，要么静态库，这样可以简化每个包的on_install过程，否则每个包就需要处理三种组合版本。

```lua
add_requires("openexr") -- 静态库
add_requires("openexr", {configs = {shared = true}}) -- 动态库
```

另外，如果确实需要一个工程里面同时需要使用静态库和动态库，add_requires也可以直接支持


```lua
add_requires("openexr~static") -- 静态库
add_requires("openexr~shared", {configs = {shared = true}}) -- 动态库
```

加上 `~static/~shared` 可以强制作为两个不同的 openexr包，这样即使 同一个 target，也可以同时使用它们。所以我觉得完全没必要 在 on_install 过程中，支持 静态库、动态库并存

Comment (xq114) [reply packages/o/openexr/xmake.lua:14]: 这个库比较特殊，它在CMakeLists.txt里面提供了同时构建选项，并且针对同时构建的情况做了特殊处理（加后缀），因此不等同于直接装两个版本的库。同时也确实有人有这个需求，见[](https://github.com/AcademySoftwareFoundation/openexr/issues/123)。从issues列表看这个选项应该属于历史遗留，为了兼容老项目，因此最好能保留。对一般的库确实不需要这个选项

Comment (waruqi) [reply packages/o/openexr/xmake.lua:14]: 哦 了解了，那就先暂时放着吧

**Pull request merged**